torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      duplicate_tv.end(),
+      [](TensorView* left, TensorView* right) {
+        return left->name() > right->name();
+      });
+  return duplicate_tv;
+}
+
+//! Find all TensorViews that require inline ComputeAt
+//! to avoid non-static allocation error
+std::vector<TensorView*> findTensorViewsToComputeAtInline(
+    Fusion* fusion,
+    const std::vector<TensorView*>& other_tv) {
+  std::vector<TensorView*> computeAt_inline_tv;
+  for (auto tv : other_tv) {
+    if (!fusion->hasInput(tv) && !fusion->hasOutput(tv)) {
+      auto domain = tv->domain()->domain();
+      bool constant_allocation = true;
+      for (size_t axis = tv->getThisComputeAtAxis(); axis < domain.size();
+           ++axis) {
+        if (!domain[axis]->isBroadcast() && !domain[axis]->isReduction()) {
+          constant_allocation &= domain[axis]->isConstScalar();
+        }
+      }
+      if (!constant_allocation) {
+        computeAt_inline_tv.push_back(tv);
+      }
+    }
+  }
+  return computeAt_inline_tv;
+}
+
+//! Place all cache TensorViews in Shared Memory
+//! All point-wise TensorViews inherit shared memory from their parents
+void setupSharedMemory(
+    Fusion* fusion,
+    const std::vector<TensorView*>& cache_tv) {
+  std::vector<TensorView*> stack(cache_tv.begin(), cache_tv.end());
+  while (!stack.empty()) {
+    auto tensor = stack.back();
+    stack.pop_back();
+    if (!fusion->hasOutput(tensor) && !fusion->hasInput(tensor)) {
+      tensor->setMemoryType(MemoryType::Shared);
+      for (auto expr : fusion->unordered_uses(tensor)) {
+        if (isPointwiseOp(expr)) {
+          auto output = expr->output(0)->as<TensorView>();
+          stack.push_back(output);
+        }
+      }
+    }
+  }
+}
+
+} // namespace
+
+void scheduleMultipleReduction(
+    Fusion* fusion,
+    const ReductionParams& rparams,
+    const std::vector<TensorView*>& reduction_tv,
+    std::vector<TensorView*>& other_tv) {
+  FusionGuard fg(fusion);
+
+  const auto& inputs_tv = ir_utils::filterByType<TensorView>(fusion->inputs());
+  const auto& outputs_tv =
+      ir_utils::filterByType<TensorView>(fusion->outputs());
+
+  // Scheduling the Reduction
+  if (rparams.fastest_dim) {
+    if (rparams.multiple_reds_per_blk) {
+      // 1) If Persistent Kernel: Automatically cache input in shared memory
+      // Else: Find TensorViews to duplicate and computeAt inline
+      if (!rparams.persistent_kernel) {
+        auto duplicate_tv = findTensorViewsToDuplicate(fusion, other_tv);
+
+        // Any TVs with multiple uses and dependencies with same IterDomain
+        // Order of Duplication is necessary for correctness
+        for (auto tensor : duplicate_tv) {
+          auto result = tensor->duplicate();
+          other_tv.insert(other_tv.end(), result.begin(), result.end());
+        }
+      }
+
+      // 2) Apply heuristics to each reduction
+      std::vector<TensorView*> rfactor_tv;
+      for (auto tv : reduction_tv) {
+        // Reduction Splits
```

Comment:
These are splits on non-reduction IterDomains but reduction tensor views correct? Could a comment syntax like Kevin used with single reduction heuristics be used here? It may help make the logic a little more clear.

Reply:
Looks like this will split both iter domains and reduction domains